experimental/coinbase-negotiator/README.md

```diff
@@ -0,0 +1,168 @@
+
+### CTV
+Using CTV you will have the above pros without the above cons so it make sense to add this
+possibility if it will be merged on master.
+
+### No coinbase negotiation
+Pool can reserve a path to itself where they do custody bitcoin for miner that prefer to use a custodian pool.
+
+### Messages
+
+#### Valid for both Schnorr and CTV negotiations:
+
+Upstream will use the provided address as receiving address for the downstream path.
+```
+Client -> Server
+NewAddress:
+    address: B0255
+    
+msg_type: 0x77
+```
+
+#### Valid for Schnorr negotiations:
+
+
+Upstream will use the pub keys as Xi and Ri to calculate X and R as described here
+https://blog.blockstream.com/en-musig-key-aggregation-schnorr-signatures/ (MuSig paragrapher)
+```
+Client -> Server
+NewPubKeyPair:
+    pub_key_x: B0255
+    pub_key_r: B0255
+
+msg_type: 0x78
+```
+
+
+
+Downstream will use X, R, m, and L to sign m as described here
+https://blog.blockstream.com/en-musig-key-aggregation-schnorr-signatures/ (MuSig paragrapher)
+```
+Server -> Client
+NewTxToSign:
+    x: B0255
+    r: B0255
+    l: B0255
+    m: B064K
+
+
+msg_type: 0x79
+```
+
+
+Upstream will use the provided si value to calculate s as described here
+https://blog.blockstream.com/en-musig-key-aggregation-schnorr-signatures/ (MuSig paragrapher)
+```
+Client -> Server
+NewSignature:
+    s: B0255
+```
+
+
+After that a tx has been signed by all the downstream, upstream compute the valid signature s and
+send it to the downstream that control the output in m
+```
+Server -> Client
+ValidSignature:
+   s: B0255
+
+
+msg_type: 0x7A
+```
+
+
+When downstream will start to mine a new block it will use the new coinbase tx if it know a valid tx
+that spend bitcoin from the coinbase output to the downstream's controlled address, and if the sum
+of all the bitcoin spent by the tx signed via NewTxToSign do not exceed the coinbase input
+```
+Server -> Client
+NewCoinbase:
+   coinbase: B064K
+
+
+msg_type: 0x7B
+```
+
+
+### Note 1 possible implementation
+Let say that a miner want to start to mine with a pool:
+1. miner will send a NewPubKeyPair to the pool.
+2. pool will not send a NewCoinbaseTx and the miner should start to mine whatever job is provided
+   by the pool.
+3. pool and miner will register all the valid shares provided by the miner.
+   [Sum(valid shares for job n), Sum(valid shares for job n+1, ...]
+4. on NewPrevHash pool send NewTxToSign where it pay to the miner all the provided shares, right now
+   the miner is still mining a job the is not paying the miner but on next block will mine a block
+   that pay the miner.
+5. from this point the pool will keep sending NewTxToSign for every NewPrevHash
+6. when miner want to exit it just stop mine new jobs and  wait until the pool find a new block. If the miner go offline pool will stop to put the miner reward in the coinbase tx but as soon as the miner will be online the miner will restart to put miner reward in coinbase.
+
```

Comment:
1. The miner will send a `NewPubKeyPair` to the Pool Service.
2. The NC Pool Service will now send a `NewCoinbaseTx` to the miner who should start mining on the job provided by the NC Pool Service.
3. The NC Pool Service and the client (downstream node) will register all the valid shares provided by the downstream Mining Devices.
`[Sum(valid shares for job n), Sum(valid shares for job n+1, ...]`
4. On a `NewPrevHash`, the NC Pool Service sends a `NewTxToSign` where it pays to the miner all the provided shares. At this same time, the miner is still mining a job. The miner does not receive the payout until the next block. 
5. From this point the NC Pool Service will keep sending `NewTxToSign`’s for every `NewPrevHash`.
6. As soon as the NC Pool Service finds a block, if miner stop accepting jobs it will be net pair with the pool.